tools/agent-prebuilt.sh

```diff
@@ -84,35 +88,51 @@ function do_clone {
             echo "Please commit these before attempting to rebase."
             return 1
         fi
-        git pull --rebase
+
+        git fetch
         if [[ $? -ne 0 ]]; then
-            echo "WARNING: Pulling from $git_url failed, which might be ok if"
-            echo "the branch the existing repository is checked out to doesn't"
-            echo "exist upstream."
+            echo "WARNING: Fetching from $git_url failed, which might be ok "
+            echo "if the branch the existing repository is checked out to "
+            echo "doesn't exist upstream."
         fi
+
         git checkout $agent_branch --
-        if [[ $? -ne 0 ]]; then
+        git_exit=$(( $git_exit + $? ))
+
+        git rebase
```

Comment:
I'm not confident that `git rebase` without args does exactly what we want here, but if you've tested it, then okay.

I would have only doing one git rebase on the final branch at the end. I think the way it is here, if we (a) *do* find a given $agent_branch, but for some reason the `git rebase` fails on it... we'll end up falling back to $branch when that might not be the correct thing to do?

Reply:
That makes sense. In fact I'm now starting to wonder if the rebase was ever doing the right thing. Before this change, when we did a git pull --rebase, that would have left us on the current branch by default, rebasing local changes onto that branch. We then checkout a completely different branch,  moving away from our local changes (!)

I think the right thing to do here is to drop the rebase support altogether - I've never used it and it's only adding complexity. Developers who need to modify agents just have to push a branch of their agent to github, set AGENT_PREBUILT_AGENT_BRANCH to the that branch name in their component Makefile, and they're done.

Reply:
Woah. Dropping the invocations of 'git rebase' resulted in us staying on the older bits from the newly checked out branch.

```
timf@jenkins-agent-x86_64-19.2.0 (prr-TOOLS-2396) ./tools/agent-prebuilt.sh  -b master -c /var/tmp/agent-cache.timf -r manta-rebalancer -u https://github.com/joyent/manta-rebalancer -p ook clone
From https://github.com/joyent/manta-rebalancer
   c92c6ab..2aba77e  engbld     -> origin/engbld
warning: unable to rmdir 'deps/eng': File exists
Switched to branch 'master'
Your branch is behind 'origin/master' by 6 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
timf@jenkins-agent-x86_64-19.2.0 (prr-TOOLS-2396)
```

whereas doing a rebase I think must imply we also do a pull on the checked-out branch, since that looks like:

```
timf@jenkins-agent-x86_64-19.2.0 (prr-TOOLS-2396) ./tools/agent-prebuilt.sh  -b master -c /var/tmp/agent-cache.timf -r manta-rebalancer -u https://github.com/joyent/manta-rebalancer -p ook clone
From https://github.com/joyent/manta-rebalancer
   c92c6ab..2aba77e  engbld     -> origin/engbld
warning: unable to rmdir 'deps/eng': File exists
Switched to branch 'master'
Your branch is behind 'origin/master' by 6 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
First, rewinding head to replay your work on top of it...
Fast-forwarded master to refs/remotes/origin/master.
timf@jenkins-agent-x86_64-19.2.0 (prr-TOOLS-2396)
```

Of note - `First, rewinding head to replay your work on top of it...
Fast-forwarded master to refs/remotes/origin/master.`

Reply:
I think simply doing a 'git pull' once we've settled on the branch we want to build is sufficient.